db/compaction/compaction_iterator.cc

```diff
@@ -639,28 +639,37 @@ void CompactionIterator::NextFromInput() {
 }
 
 void CompactionIterator::PrepareOutput() {
-  // Zeroing out the sequence number leads to better compression.
-  // If this is the bottommost level (no files in lower levels)
-  // and the earliest snapshot is larger than this seqno
-  // and the userkey differs from the last userkey in compaction
-  // then we can squash the seqno to zero.
-  //
-  // This is safe for TransactionDB write-conflict checking since transactions
-  // only care about sequence number larger than any active snapshots.
-  //
-  // Can we do the same for levels above bottom level as long as
-  // KeyNotExistsBeyondOutputLevel() return true?
-  if ((compaction_ != nullptr && !compaction_->allow_ingest_behind()) &&
-      ikeyNotNeededForIncrementalSnapshot() && bottommost_level_ && valid_ &&
-      IN_EARLIEST_SNAPSHOT(ikey_.sequence) && ikey_.type != kTypeMerge) {
-    assert(ikey_.type != kTypeDeletion && ikey_.type != kTypeSingleDeletion);
-    if (ikey_.type == kTypeDeletion || ikey_.type == kTypeSingleDeletion) {
-      ROCKS_LOG_FATAL(info_log_,
-                      "Unexpected key type %d for seq-zero optimization",
-                      ikey_.type);
+  if (valid_) {
+    if (compaction_filter_ && ikey_.type == kTypeBlobIndex) {
+      if (compaction_filter_->PrepareBlobOutput(user_key(), value_,
+                                                &compaction_filter_value_)) {
+        value_ = compaction_filter_value_;
```

Comment:
If my understanding is correct, any blob read/write failure will end up with not GCing but leaving the blob files in original places. Is it correct? Either way, we should add a comment to explain the failure cases.

I do prefer handling it in a panic way which fail the compaction as a whole to make it more visible though, but it's debatable.

Reply:
Yes, currently any read/write failure results in the blob index not being updated. I was actually thinking about doing the "all or nothing" approach you mentioned; will look into what that would entail.